### PR TITLE
feat (core): support converting attachments to file parts

### PR DESCRIPTION
**Pull request by jeremyphilemon**
- [x] if url is provided, and attachment is not image, then convert to file part
- [x] if data url is provided, and attachment is not image, then convert to file part
- [x] throw error when file attachment does not specify `contentType`

### CODE REVIEW

File: packages/ai/core/prompt/attachments-to-parts.ts [hunk 1]
```diff
@@ -61,6 +73,18 @@ export function attachmentsToParts(attachments: Attachment[]): ContentPart[] {
...
+        } else {
+          if (!attachment.contentType) {
+            throw new Error(
+              'If the attachment is not an image or text, it must specify a content type',
+            );
+          }
+
+          parts.push({
+            type: 'file',
+            data: convertDataContentToUint8Array(base64Content),
```
Comment (lgrammel) [inline packages/ai/core/prompt/attachments-to-parts.ts:85]: this is unnecessary (also in the older code) and i saw hints that we use up cpu a lot. you can just pass in the base64

File: packages/ai/core/prompt/convert-to-core-messages.test.ts [hunk 2]
```diff
@@ -87,6 +87,24 @@ describe('user message', () => {
...
+        {
+          role: 'user',
+          content: 'Check this file',
+          experimental_attachments: [attachment],
+        },
+      ]);
+    }).toThrow(
+      'If the attachment is not an image or text, it must specify a content type',
+    );
+  });
```
Comment (lgrammel) [inline packages/ai/core/prompt/convert-to-core-messages.test.ts:106]: test happy path as well?

**Review by lgrammel — APPROVED** [review]

File: packages/ai/core/prompt/convert-to-core-messages.test.ts (hunk 2 above)
Comment (jeremyphilemon) [reply packages/ai/core/prompt/convert-to-core-messages.test.ts:164]: added!

**Comment by xsayedmahmud** [comment]
Does this work yet, i updated to version 3.4.9 and i am still encountering problems with any other files except images. 

**File Handling Process:**
1. For all file types (except images):
   - I extract the content from files and save it as text fromat in supabase.
   - I use the Supabase URL of the saved content in the message.
   - I am converting to text files because at first i thought only text and image files are handled. But later found about this merge to support any file format. 

2. For images:
   - I upload the image directly to Supabase.
   - I use the Supabase URL of the uploaded image in the message.

3. Message format for my chat history looks like this:
   ```
{
        "id": "ALFOFyI",
        "createdAt": "2024-10-06T23:31:07.587Z",
        "role": "user",
        "content": "hi"
    },
    {
        "id": "LLwT9Vr",
        "role": "assistant",
        "content": "Hello again! How are you today? Is there something specific you would like to talk about or ask me? I'm here to help!",
        "createdAt": "2024-10-06T23:31:10.617Z"
    },
    {
        "id": "JcqmYJx",
        "createdAt": "2024-10-06T23:35:41.441Z",
        "role": "user",
        "content": "what is this file about",
        "experimental_attachments": [
            {
                "name": "profile.pdf",
                "contentType": "text/plain",
                "url": "https://lmuyighwbqqhdskfwlsr.supabase.co/storage/v1/object/public/text_files/7xxx8d0-0cc9-4cf0-af8c-660ce76exxxf.txt"
            }
        ]
    }
   ```
4. The error message i see: 

**Error Message:**
```
{
  "detail":[
    {
      "type":"string_type",
      "loc":["body","messages",4,"content","str"],
      "msg":"Input should be a valid string",
      "input":[{"type":"text","text":"what is this file about"},null]
    },
   
  ]
}
```

**Additional Information:**
- This issue occurs consistently with various non-image file types.
- Messages with text-only content or image attachments are processed correctly.
- The error suggests that the API is expecting a string for the message content, but is receiving something else when a non-image file is attached.


 5. Complete error message i see: 

```
cause: undefined,
  url: 'https://api.hyperbolic.xyz/v1/chat/completions',
  requestBodyValues: {
    model: 'mistralai/Pixtral-12B-2409',
    logit_bias: undefined,
    logprobs: undefined,
    top_logprobs: undefined,
    user: undefined,
    parallel_tool_calls: undefined,
    max_tokens: undefined,
    temperature: 1,
    top_p: undefined,
    frequency_penalty: undefined,
    presence_penalty: undefined,
    stop: undefined,
    seed: undefined,
    response_format: undefined,
    messages: [ [Object], [Object], [Object] ],
    tools: undefined,
    tool_choice: undefined,
    stream: true,
    stream_options: undefined
  },
  statusCode: 422,
  responseHeaders: {
    'cf-cache-status': 'DYNAMIC',
    'cf-ray': '8ce994713e7e789b-JED',
    connection: 'keep-alive',
    'content-length': '506',
    'content-type': 'application/json',
    date: 'Sun, 06 Oct 2024 23:55:01 GMT',
    server: 'cloudflare'
  },
  responseBody: '{"detail":[{"type":"string_type","loc":["body","messages",2,"content","str"],"msg":"Input should be a valid string","input":[{"type":"text","text":"what is this file about?"},null]},{"type":"dict_type","loc":["body","messages",2,"content","list[union[typed-dict,typed-dict]]",1,"typed-dict"],"msg":"Input should be a valid dictionary","input":null},{"type":"dict_type","loc":["body","messages",2,"content","list[union[typed-dict,typed-dict]]",1,"typed-dict"],"msg":"Input should be a valid dictionary","input":null}]}',
  isRetryable: false,
  data: undefined,
  [Symbol(vercel.ai.error)]: true,
  [Symbol(vercel.ai.error.AI_APICallError)]: true
}
 POST /api/chat 500 in 1069ms
```

*I tried with other providers, like openai, mistral, cohere, they all throws similar error.

**Comment by xsayedmahmud** [comment]
@jeremyphilemon @lgrammel Could you please look into the file handling issue I detailed in my comment above? I still coudn't resolve it.

**Comment by xsayedmahmud** [comment]
@jeremyphilemon @lgrammel Could you please review the file handling issue I described in my previous comment to see if I'm making any mistakes or suggest how to resolve the problem?

**Pull request merged**